rest-client/src/main/java/org/sagebionetworks/bridge/rest/RestUtils.java

```diff
@@ -198,20 +198,47 @@ public static String getUserAgent(ClientInfo info) {
         if (info == null) {
             return null;
         }
-        List<String> stanzas = Lists.newArrayListWithCapacity(3);
-        if (isNotBlank(info.getAppName()) && info.getAppVersion() != null) {
-            stanzas.add(String.format("%s/%s", info.getAppName(), info.getAppVersion()));
+        // Send what is available. The server can handle all meaningful combinations.
+        StringBuilder sb = new StringBuilder();
+        if (isNotBlank(info.getAppName()) || info.getAppVersion() != null) {
+            if (isNotBlank(info.getAppName())) {
+                sb.append(info.getAppName());
+            }
+            if (info.getAppVersion() != null) {
+                sb.append("/");
+                sb.append(info.getAppVersion());
+            }
         }
-        if (isNotBlank(info.getDeviceName()) && isNotBlank(info.getOsName()) && isNotBlank(info.getOsVersion())) {
-            stanzas.add(String.format("(%s; %s/%s)", info.getDeviceName(), info.getOsName(), info.getOsVersion()));
+        if (isNotBlank(info.getDeviceName()) || isNotBlank(info.getOsName()) || isNotBlank(info.getOsVersion())) {
+            sb.append(" (");
+            // Only include device name if there's also os, this is one ambiguity we cannot parse.
+            if (isNotBlank(info.getDeviceName())) {
+                sb.append(info.getDeviceName());
+                sb.append("; ");
+            }
+            if (isNotBlank(info.getOsName())) {
+                sb.append(info.getOsName());
+            }
+            if (isNotBlank(info.getOsVersion())) {
+                sb.append("/");
+                sb.append(info.getOsVersion());
+            }
+            sb.append(")");
         }
-        if (!stanzas.isEmpty() && isNotBlank(info.getSdkName()) && info.getSdkVersion() != null) {
-            stanzas.add(String.format("%s/%s", info.getSdkName(), info.getSdkVersion()));    
+        if (isNotBlank(info.getSdkName()) || info.getSdkVersion() != null) {
+            sb.append(" ");
+            if (isNotBlank(info.getSdkName())) {
+                sb.append(info.getSdkName());
+            }
+            if (info.getSdkVersion() != null) {
+                sb.append("/");
+                sb.append(info.getSdkVersion());
+            }
```

Comment:
Similarly, tests for SDK name without version and SDK version without name.

Reply:
Added a bunch of new tests for a bunch of combinations that'll now work